docs/arch/adr-010-new-versioning-schema.md

```diff
@@ -0,0 +1,145 @@
+
+| Add-on | Version |
+|----------|-------|
+|Kubernetes| v1.26.1|
+|Containerd| v1.6.3 |
+|Flannel | v0.21.4|
+|Contour | v1.24.3 |
+|Minio | 2023-04-13T03-08-07Z |
+|OpenEBS | v3.5.0 |
+|Rook | v1.11.3 |
+
+It's important to note that the Kubernetes minor version remains unchanged within the same Channel. Any upgrades to add-on versions must be compatible with their first kURL version in the Channel (like for example: upgrading from OpenEBS 3.4.0 to 3.5.0 can be achieved without data migration). This may sometimes involve _minor_ version upgrades, but in most cases, only _patch_ version upgrades will be allowed within a Channel.
+
+### Upgrades between Channels
+
+Upgrades between different Channels can possibly involve longer cluster downtime due to data migration. This may become very complex as some add-ons may not be supported in updated Kubernetes versions. Therefore, it's crucial for us to keep track of the current state of the cluster before staring the upgrade.
+
+To manage upgrades between Channels, we have decided to allow only one Channel version upgrade at a time. This means that customers can upgrade from Channel 1.25 to 1.26, but upgrades from 1.25 to 1.27 would be blocked. Additionally, users must be in the **latest** version within one Channel before upgrading to any version in the next Channel. By requiring users to be in the latest Channel version before upgrading to the next one, we can ensure two things: Firstly, we have a clear understanding of the current state of the cluster, which is essential for successful upgrades. Secondly, it provides us with some flexibility in preparing the cluster for the Channel upgrade, as we can introduce a new version in the previous Channel to gradually prepare the cluster to the next Channel.
+
+To help us better comprehend the restrictions and limitations involved in upgrading between Channels, let us consider a hypothetical scenario. This will provide us with a practical example to illustrate the concept and highlight the challenges that may arise during the upgrading process.
+
+| Channel | Version|
+|---------|--------|
+| 1.25    | 1.0.0  |
+| 1.25    | 1.1.0  |
+| 1.25    | 1.2.0  |
+| 1.26    | 1.0.0  |
+| 1.26    | 1.1.0  |
+| 1.26    | 1.2.0  |
+| 1.27    | 1.0.0  |
+
+Based on the above-mentioned scenarios, the following table outlines various upgrade attempts and their respective outcomes:
+
+| From       | To        | Outcome |
+|------------|-----------|---------|
+|v1.25+1.0.0.|v1.25+1.2.0|Success. Upgrades inside the same Channel are allowed. |
+|v1.25+1.0.0.|v1.26+1.0.0|Failure. Users can only upgrade to the next Channel when they are already running the latest version of the current Channel. |
+|v1.25+1.2.0.|v1.26+1.2.0|Success. Upgrades from the latest version of the current Channel to any version on the next Channel are allowed. |
```

Comment:
I question whether we should allow upgrades to anything but the latest version in a channel. One thing is for sure we should not allow downgrading the kurl version. The question above impacts this.

Reply:
That is indeed valid of a note. I think it is a valid idea.

Reply:
They way I was seeing this is (and I might have just now realised what you mean by pinning the kURL version) :

1) the cluster has a version.
2) the binary as a version.

Binary version `v1.27+1.2.0` installs or upgrades the current cluster to the version 1.2.0 on for Channel 1.27. Binary and configuration are two distinct things. The configuration only indicates what binary should be downloaded and is used by the binary to assess what needs to be enabled/disabled.